portalnet/src/config.rs

```diff
@@ -0,0 +1,56 @@
+use std::net::SocketAddr;
+
+use ethereum_types::H256;
+
+use crate::metrics::overlay::OverlayMetrics;
+use crate::metrics::portalnet::PORTALNET_METRICS;
+use ethportal_api::types::bootnodes::Bootnodes;
+use ethportal_api::types::cli::TrinConfig;
+use ethportal_api::types::distance::Distance;
+
+/// Capacity of the cache for observed `NodeAddress` values.
+/// Provides capacity for 32 full k-buckets. This capacity will be shared among all active portal
+/// subnetworks.
+const NODE_ADDR_CACHE_CAPACITY: usize = discv5::kbucket::MAX_NODES_PER_BUCKET * 32;
+
+#[derive(Clone)]
+pub struct PortalnetConfig {
+    pub external_addr: Option<SocketAddr>,
+    pub private_key: H256,
+    pub listen_port: u16,
+    pub bootnodes: Bootnodes,
+    pub data_radius: Distance,
+    pub internal_ip: bool,
+    pub no_stun: bool,
+    pub node_addr_cache_capacity: usize,
+    pub metrics: OverlayMetrics,
+}
+
+impl PortalnetConfig {
+    pub fn new(trin_config: &TrinConfig, private_key: H256) -> Self {
+        Self {
+            external_addr: trin_config.external_addr,
+            private_key,
+            listen_port: trin_config.discovery_port,
+            no_stun: trin_config.no_stun,
+            bootnodes: trin_config.bootnodes.clone(),
+            ..Default::default()
+        }
+    }
+}
+
+impl Default for PortalnetConfig {
+    fn default() -> Self {
+        Self {
+            external_addr: None,
+            private_key: H256::random(),
+            listen_port: 4242,
+            bootnodes: Bootnodes::default(),
+            data_radius: Distance::MAX,
+            internal_ip: false,
+            no_stun: false,
+            node_addr_cache_capacity: NODE_ADDR_CACHE_CAPACITY,
+            metrics: PORTALNET_METRICS.overlay.clone(),
```

Comment:
I don't understand why the metrics are inside the config here. It doesn't really seem to belong with the other configured portal network values (simple entries from the CLI). I think the whole PR gets simpler, and the code makes more sense, if the config is left alone and the metrics are just handled separately.

Reply:
Yeah, good call. That design was cruft leftover from a previous design where I manually created the metrics registry and passed it through the configs to the respective services. Now using the `lazy_static` to globally initialize eliminates the need for this.